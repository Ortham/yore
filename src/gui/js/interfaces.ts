export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface Photo {
  path: string;
  location?: Location; // eslint-disable-line no-restricted-globals
  error?: string;
  height?: number;
  width?: number;
  loaded?: boolean;
}

interface Location {
  Existing?: Coordinates;
  Suggested?: [Coordinates, LocationAccuracy];
}

export interface LocationAccuracy {
  meters: number;
  seconds: number;
}

export interface LocationResponse {
  location?: Location; // eslint-disable-line no-restricted-globals
  error?: string;
}
