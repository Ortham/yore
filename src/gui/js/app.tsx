import '@babel/polyfill';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import 'whatwg-fetch';
import { Page, PageProps } from './page';
import {
  getInterpolate,
  getLocationHistoryPath,
  getPhotos,
  getRootPath,
  getLocations
} from './requests';
import { updatePhotoLocations } from './photo';

function getInitialState() {
  const state: PageProps = {
    interpolate: false,
    locationHistoryPath: undefined,
    photos: [],
    rootPath: undefined
  };

  return getRootPath()
    .then(responseBody => {
      state.rootPath = responseBody.rootPath || '';

      return getPhotos();
    })
    .then(photos =>
      getLocations(0, photos.length).then(locations =>
        updatePhotoLocations(photos, locations)
      )
    )
    .then(photos => {
      state.photos = photos;

      return getInterpolate();
    })
    .then(responseBody => {
      state.interpolate = responseBody.interpolate;

      return getLocationHistoryPath();
    })
    .then(responseBody => {
      state.locationHistoryPath = responseBody.locationHistoryPath || '';

      return state;
    });
}

getInitialState().then(state => {
  ReactDOM.render(<Page {...state} />, document.getElementById('root'));
});
