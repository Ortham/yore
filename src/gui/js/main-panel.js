import React from 'react';
import MapArea from './map-area';
import {locationDescription, hasSuggestion} from './photo';

export default function MainPanel(props) {
    if (props.photo) {
        return (
            <main>
                <section>
                    <img src={props.photo.src} />
                    <MapArea photo={props.photo} />
                </section>
                <footer>
                    {locationDescription(props.photo)}
                    <br />
                    <button
                        disabled={!hasSuggestion(props.photo)}
                        onClick={props.handleSuggestionApply} >
                        Apply
                    </button>
                    <button
                        disabled={!hasSuggestion(props.photo)}
                        onClick={props.handleSuggestionDiscard} >
                        Discard
                    </button>
                </footer>
            </main>
        );
    } else {
        return (
            <main>
                <section>No photo selected.</section>
                <footer>
                    <br />
                    <button disabled>Apply</button>
                    <button disabled>Discard</button>
                </footer>
            </main>
        );
    }
}
