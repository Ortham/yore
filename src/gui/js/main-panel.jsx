import React from 'react';
import PropTypes from 'prop-types';
import MapArea from './map-area';
import { locationDescription, hasSuggestion } from './photo';

export default function MainPanel(props) {
  if (props.photo) {
    return (
      <div id="main">
        <section>
          <img src={props.photo.src} alt="Selected" />
          <MapArea photo={props.photo} />
        </section>
        <footer>
          <div>{locationDescription(props.photo)}</div>
          <div>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionApply}
            >
              Apply
            </button>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionDiscard}
            >
              Discard
            </button>
          </div>
        </footer>
      </div>
    );
  }
  return (
    <div id="main">
      <section>No photo selected.</section>
      <footer>
        <div />
        <div>
          <button disabled>Apply</button>
          <button disabled>Discard</button>
        </div>
      </footer>
    </div>
  );
}

MainPanel.propTypes = {
  photo: PropTypes.shape({
    src: PropTypes.string
  }),
  handleSuggestionApply: PropTypes.func.isRequired,
  handleSuggestionDiscard: PropTypes.func.isRequired
};

MainPanel.defaultProps = {
  photo: undefined
};
