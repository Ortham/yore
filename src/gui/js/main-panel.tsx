import * as React from 'react';
import { Photo } from './interfaces';
import { MapArea } from './map-area';
import { hasSuggestion, locationDescription } from './photo';

export interface MainPanelProps {
  photo?: Photo;
  handleSuggestionApply: () => Promise<void>;
  handleSuggestionDiscard: () => void;
}

export function MainPanel(props: MainPanelProps) {
  if (props.photo) {
    return (
      <div id="main">
        <section>
          <img src={props.photo.src} alt="Selected" />
          <MapArea photo={props.photo} />
        </section>
        <footer>
          <div>{locationDescription(props.photo)}</div>
          <div>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionApply}
            >
              Apply
            </button>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionDiscard}
            >
              Discard
            </button>
          </div>
        </footer>
      </div>
    );
  }
  return (
    <div id="main">
      <section>No photo selected.</section>
      <footer>
        <div />
        <div>
          <button disabled>Apply</button>
          <button disabled>Discard</button>
        </div>
      </footer>
    </div>
  );
}
