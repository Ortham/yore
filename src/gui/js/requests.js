function get(url) {
  return fetch(url).then(response => {
    if (response.ok) {
      return response.json();
    }
    throw new Error(`Failed to GET ${url}`);
  });
}

function put(url, body) {
  const init = {
    method: 'PUT',
    body: JSON.stringify(body)
  };

  return fetch(url, init).then(response => {
    if (!response.ok) {
      throw new Error(`Failed to ${init.method} ${url} with body ${init.body}`);
    }
  });
}

function setPhotoSrc(photo) {
  return Object.assign(photo, {
    src: `/thumbnail?path=${encodeURIComponent(
      photo.path
    )}&maxWidth=500&maxHeight=500`
  });
}

function mapPhotos(responseBody) {
  return responseBody.photos.map(setPhotoSrc);
}

export function getRootPath() {
  return get('/rootPath');
}

export function getPhotos() {
  return get('/photos').then(mapPhotos);
}

export function getFilteredPhotos() {
  return get('/photos?filter').then(mapPhotos);
}

export function writeCoordinates(path, coordinates) {
  return put(`/location?path=${encodeURIComponent(path)}`, coordinates);
}

export function getLocations(startIndex, endIndex) {
  return get(`/locations?start=${startIndex}&end=${endIndex}`).then(
    responseBody => responseBody.locations
  );
}

export function getLocation(path) {
  return get(`/location?path=${encodeURIComponent(path)}`);
}
