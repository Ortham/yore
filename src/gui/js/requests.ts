import { Coordinates } from './interfaces';

function get(url: string) {
  return fetch(url).then(response => {
    if (response.ok) {
      return response.json();
    }
    throw new Error(`Failed to GET ${url}`);
  });
}

function put(url: string, body: any) {
  const init = {
    body: JSON.stringify(body),
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json'
    }
  };

  return fetch(url, init).then(response => {
    if (!response.ok) {
      throw new Error(`Failed to ${init.method} ${url} with body ${init.body}`);
    }
  });
}

export function getRootPath() {
  return get('/rootPath');
}

export function getNewRootPath() {
  return get('/rootPath/new');
}

export function getLocationHistoryPath() {
  return get('/locationHistoryPath');
}

export function getNewLocationHistory() {
  return get('/locationHistory/new');
}

export function getInterpolate() {
  return get('/interpolate');
}

export function putInterpolate(interpolate: boolean) {
  return put('/interpolate', { interpolate });
}

export function getPhotos() {
  return get('/photos').then(body => body.photos);
}

export function getFilteredPhotos() {
  return get('/photos?filter=true').then(body => body.photos);
}

export function writeCoordinates(path: string, coordinates: Coordinates) {
  return put(`/location?path=${encodeURIComponent(path)}`, coordinates);
}

export function getLocations(startIndex: number, endIndex: number) {
  return get(`/locations?start=${startIndex}&end=${endIndex}`).then(
    responseBody => responseBody.locations
  );
}

export function getLocation(path: string) {
  return get(`/location?path=${encodeURIComponent(path)}`);
}
