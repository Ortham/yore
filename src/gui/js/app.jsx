import 'babel-polyfill';
import 'whatwg-fetch';
import React from 'react';
import ReactDOM from 'react-dom';
import Page from './page';
import { getPhotos, getRootPath } from './requests';

function getInitialState() {
  const state = {
    rootPath: undefined,
    filterPhotos: false,
    currentPhoto: undefined,
    photos: []
  };

  return getRootPath()
    .then(responseBody => {
      state.rootPath = responseBody.rootPath;

      return getPhotos();
    })
    .then(photos => {
      state.photos = photos;

      return state;
    });
}

getInitialState().then(state => {
  ReactDOM.render(
    <Page rootPath={state.rootPath} photos={state.photos} />,
    document.getElementById('root')
  );
});
