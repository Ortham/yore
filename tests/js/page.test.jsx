import React from 'react';
import renderer from 'react-test-renderer';

jest.mock('../../src/gui/js/main-panel', () => 'MainPanel');
jest.mock('../../src/gui/js/sidebar', () => 'Sidebar');
jest.unmock('../../src/gui/js/requests');

import * as requests from '../../src/gui/js/requests'; // eslint-disable-line import/first
import Page from '../../src/gui/js/page'; // eslint-disable-line import/first

describe('Page', () => {
  let page;

  beforeAll(() => {
    const photos = [
      {
        height: 5,
        width: 10,
        loaded: false,
        path: 'path 1',
        src: 'source 1'
      },
      {
        height: 5,
        width: 10,
        loaded: true,
        path: 'path 2',
        src: 'source 2',
        location: {
          Suggested: [
            {
              latitude: 52.0,
              longitude: 13.2
            },
            {
              meters: 5,
              seconds: 20
            }
          ]
        }
      }
    ];

    requests.writeCoordinates = jest
      .fn()
      .mockReturnValueOnce(Promise.resolve());
    requests.getFilteredPhotos = jest
      .fn()
      .mockReturnValueOnce(Promise.resolve([photos[1]]));
    requests.getLocation = jest.fn().mockReturnValue(Promise.resolve());
    requests.getLocations = jest.fn().mockReturnValue(
      Promise.resolve([
        { error: 'Oh no!' },
        {
          location: {
            Existing: {
              latitude: 5,
              longitude: 30
            }
          }
        }
      ])
    );

    page = renderer.create(<Page rootPath="" photos={photos} />, {
      createNodeMock: element => {
        if (element.type === 'Sidebar') {
          return {
            forceUpdate: jest.fn()
          };
        }
        return null;
      }
    });
  });

  beforeEach(() => {
    requests.writeCoordinates.mockClear();
    requests.getFilteredPhotos.mockClear();
    requests.getLocation.mockClear();
    requests.getLocations.mockClear();
  });

  test('renders a header, sidebar and main panel', () => {
    expect(page.toJSON()).toMatchSnapshot();
  });

  test('handlePhotoSelect should set the current photo to the given photo', () => {
    const pageInstance = page.root.instance;

    expect(pageInstance.state.currentPhoto).toBe(undefined);

    pageInstance.handlePhotoSelect(pageInstance.state.photos[1]);

    expect(pageInstance.state.currentPhoto).toEqual(
      pageInstance.state.photos[1]
    );
    expect(pageInstance.sidebar.forceUpdate.mock.calls.length).toBe(1);
  });

  test('handleSuggestionApply calls writeCoordinates then moves suggested location to existing', () => {
    const pageInstance = page.root.instance;

    pageInstance.handlePhotoSelect(pageInstance.state.photos[1]);

    return pageInstance.handleSuggestionApply().then(() => {
      expect(requests.writeCoordinates.mock.calls.length).toBe(1);
      expect(pageInstance.state.currentPhoto.location).toEqual({
        Existing: {
          latitude: 52.0,
          longitude: 13.2
        }
      });
      expect(pageInstance.state.currentPhoto).toEqual(
        pageInstance.state.photos[1]
      );
      expect(pageInstance.sidebar.forceUpdate.mock.calls.length).toBe(1);
    });
  });

  test("handleSuggestionDiscard should set the current photo's location to be undefined", () => {
    const pageInstance = page.root.instance;

    pageInstance.handlePhotoSelect(pageInstance.state.photos[1]);

    pageInstance.handleSuggestionDiscard();

    expect(pageInstance.state.currentPhoto.location).toBe(undefined);
    expect(pageInstance.state.currentPhoto).toEqual(
      pageInstance.state.photos[1]
    );
    expect(pageInstance.sidebar.forceUpdate.mock.calls.length).toBe(1);
  });

  test('handleFilterToggle should call getFilteredPhotos if the filter is enabled', () => {
    const pageInstance = page.root.instance;

    expect(pageInstance.state.filterPhotos).toBe(false);

    return pageInstance
      .handleFilterToggle({ target: { checked: true } })
      .then(() => {
        expect(requests.getFilteredPhotos.mock.calls.length).toBe(1);
        expect(pageInstance.state.filterPhotos).toBe(true);
        expect(pageInstance.state.sidebarPhotos.length).toBe(1);
        expect(pageInstance.state.sidebarPhotos[0]).toEqual(
          pageInstance.state.photos[1]
        );
      });
  });

  test('handleFilterToggle should call set full photos array if the filter is disabled', () => {
    const pageInstance = page.root.instance;

    return pageInstance
      .handleFilterToggle({ target: { checked: false } })
      .then(() => {
        expect(requests.getFilteredPhotos.mock.calls.length).toBe(0);
        expect(pageInstance.state.filterPhotos).toBe(false);
        expect(pageInstance.state.sidebarPhotos.length).toBe(2);
        expect(pageInstance.state.sidebarPhotos).toEqual(
          pageInstance.state.photos
        );
      });
  });

  test('getLocationsPromise calls getLocation for each photo if filterPhotos is true', () => {
    const pageInstance = page.root.instance;

    pageInstance.setState(
      Object.assign({}, pageInstance.state, { filterPhotos: true })
    );

    return pageInstance.getLocationsPromise(0, 2).then(() => {
      expect(requests.getLocation.mock.calls.length).toBe(2);
    });
  });

  test('getLocationsPromise calls getLocations for range if filterPhotos is false', () => {
    const pageInstance = page.root.instance;

    pageInstance.setState(
      Object.assign({}, pageInstance.state, { filterPhotos: false })
    );

    return pageInstance.getLocationsPromise(0, 2).then(() => {
      expect(requests.getLocations.mock.calls.length).toBe(1);
    });
  });

  test('getAndStoreLocations sets location, error and loaded photo fields', () => {
    const pageInstance = page.root.instance;

    return pageInstance.getAndStoreLocations(0, 2).then(() => {
      expect(requests.getLocations.mock.calls.length).toBe(1);

      expect(pageInstance.state.sidebarPhotos[0].location).toBe(undefined);
      expect(pageInstance.state.sidebarPhotos[0].error).toBe('Oh no!');
      expect(pageInstance.state.sidebarPhotos[0].loaded).toBe(true);
      expect(pageInstance.state.sidebarPhotos[1].location).toEqual({
        Existing: {
          latitude: 5,
          longitude: 30
        }
      });
      expect(pageInstance.state.sidebarPhotos[1].error).toBe(undefined);
      expect(pageInstance.state.sidebarPhotos[1].loaded).toBe(true);
    });
  });
});
