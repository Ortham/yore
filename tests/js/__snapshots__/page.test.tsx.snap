// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page renders a header, photos grid and main panel 1`] = `
<div>
  <header
    id="titleBar"
  >
    <h1>
      Yore
    </h1>
    <div
      id="paths"
    >
      <div>
        Photos path: 
        
      </div>
      <div>
        Location history path: 
        
      </div>
    </div>
    <div>
      <div>
        <button
          onClick={[Function]}
        >
          Select Root Path
        </button>
        <label
          htmlFor="suggestionsCheckbox"
        >
          <input
            checked={false}
            id="suggestionsCheckbox"
            onChange={[Function]}
            type="checkbox"
          />
          Show only photos with suggestions
        </label>
      </div>
      <div>
        <button
          onClick={[Function]}
        >
          Select Location History
        </button>
        <label
          htmlFor="interpolateCheckbox"
        >
          <input
            checked={false}
            id="interpolateCheckbox"
            onChange={[Function]}
            type="checkbox"
          />
          Interpolate locations
        </label>
      </div>
    </div>
  </header>
  <div>
    <PhotosGrid
      currentPhoto={undefined}
      getAndStoreLocations={[Function]}
      handlePhotoSelect={[Function]}
      photos={
        Array [
          Object {
            "height": 5,
            "loaded": false,
            "path": "path 1",
            "src": "source 1",
            "width": 10,
          },
          Object {
            "height": 5,
            "loaded": true,
            "location": Object {
              "Suggested": Array [
                Object {
                  "latitude": 52,
                  "longitude": 13.2,
                },
                Object {
                  "meters": 5,
                  "seconds": 20,
                },
              ],
            },
            "path": "path 2",
            "src": "source 2",
            "width": 10,
          },
        ]
      }
    />
    <MainPanel
      handleSuggestionApply={[Function]}
      handleSuggestionDiscard={[Function]}
      photo={undefined}
    />
  </div>
</div>
`;
